File: 0514-Freedom Trail/0514-Freedom Trail.cpp
class Solution {
public:
    int findRotateSteps(string ring, string key) {
        int n = ring.size();
        
        unordered_map<char, vector<int>> table;
        for (int i = 0; i < n; ++i) {
            table[ring[i]].push_back(i);
        }
        
        int m = key.size();
        vector<vector<int>> dp(m, vector<int>(n, INT_MAX));
        for (int j = 0; j < n; ++j) {
            if (key[0] == ring[j]) {
                dp[0][j] = min(j, n - j) + 1;
            }
        }

        for (int i = 1; i < m; ++i) {
            for (int j : table[key[i]]) {
                for (int k : table[key[i - 1]]) {
                    dp[i][j] = min(dp[i][j], dp[i - 1][k] + min(abs(j - k), n - abs(j - k)) + 1);
                }
            }
        }
        
        int result = INT_MAX;
        for (int j : table[key[m - 1]]) {
            result = min(result, dp[m - 1][j]);
        }
        
        return result;
    }
};

File: 0501-0600/0514-Freedom Trail/0514-Freedom Trail.py
import math
import collections
class Solution:
    def findRotateSteps(self, ring: str, key: str) -> int:
        table = collections.defaultdict(list)
        for i, c in enumerate(ring):
            table[c].append(i)

        dp = [[math.inf]*len(ring) for _ in range(len(key))]
        for j in range(len(ring)):
            if key[0] == ring[j]:
                dp[0][j] = min(j, len(ring) - j) + 1
        
        for i in range(1, len(key)):
            for j in table[key[i]]:
                for k in table[key[i - 1]]:
                    dp[i][j] = min(dp[i][j], dp[i - 1][k] + min(abs(j - k), len(ring) - abs(j - k)) + 1)
                    
        result = math.inf
        for j in table[key[-1]]:
            result = min(result, dp[-1][j])
        return result
